### PR TITLE
Fixes from CSAF validation

### PR DESCRIPTION
**Pull request by sustefil**
CSAF validation triggers a need for some minor fixes, changing types, creating mapping dicts etc.

### CODE REVIEW

File: cvrf2csaf/section_handlers/product_tree.py [hunk 1]
```diff
@@ -4,7 +4,22 @@
...
+        "Product Version": 'product_version',
+        "Patch Level": 'patch_level',
+        "Service Pack": 'service_pack',
+        "Architecture": 'architecture',
+        "Language": 'language',
+        "Legacy": 'legacy',
+        "Specification": 'specification',
+        "Host Name": 'host_name',
+        "Realm": 'N/A',
+        "Resource": 'N/A',
```
Comment (sustefil) [inline cvrf2csaf/section_handlers/product_tree.py:20]: Hi @tschmidtb51 

1) what to do with these 2 values: `Realm` and `Resource`? They are valid in the input, but have no mapping to the output

2) In general, do you agree with creating these mapping dicts for string values? I like to be explicit and I find this more readable than having .lower() somewhere deeper in the code. Furthermore, sometimes these transformations (like str.lower), aren't enough.

**Review by tschmidtb51 — COMMENTED** [review]
Please use [strict schema](https://github.com/secvisogram/secvisogram/blob/main/app/lib/shared/Core/csaf_2.0_strict.json) instead. This is analyzing the header fields (keys). Otherwise invalid variable names are not throwing a error. strict version is also used in secvisogram.

File: cvrf2csaf/section_handlers/product_tree.py (hunk 1 above)
Comment (cgi1) [reply cvrf2csaf/section_handlers/product_tree.py:20]: there is a conversion rule, Thomas will update this here.

File: cvrf2csaf/config/config.yaml [hunk 2]
```diff
@@ -1,13 +1,13 @@
 # TODO: use version derived from git revision
-cvrf2csaf_version: 1.0
+cvrf2csaf_version: "1.0"
```
[the remark this answers is not in the document]
Comment (cgi1) [reply cvrf2csaf/config/config.yaml:2]: @sustefil: Please check how to get the version number from the package (e.g. by a plugin as used in secvisogram called `gitRevisionPlugin`)

File: cvrf2csaf/section_handlers/product_tree.py (hunk 1 above)
Comment (tschmidtb51) [reply cvrf2csaf/section_handlers/product_tree.py:20]: **Conversion rule**: Convert them into a `product_name` and add a warning that those types do not exist in CSAF.

File: cvrf2csaf/cvrf2csaf.py [hunk 3]
```diff
@@ -74,16 +74,24 @@ def _compose_final_csaf(self) -> dict:
...
-            final_csaf['document']['acknowledgments'] = self.document_acknowledgments.csaf
-        final_csaf['document']['notes'] = self.document_notes.csaf
         final_csaf['document']['publisher'] = self.document_publisher.csaf
         final_csaf['document']['tracking'] = self.document_tracking.csaf
-        final_csaf['document']['references'] = self.document_references.csaf
-        final_csaf['product_tree'] = self.product_tree.csaf
-        final_csaf['vulnerabilities'] = self.vulnerability.csaf
+
+        # Optional root elements
+        if self.document_notes.csaf:
```
Comment (sthagen) [inline cvrf2csaf/cvrf2csaf.py:84]: A more data centered approach might do this:
```python
        mappings = (
            (final_csaf['document'], 'notes, self.document_notes.csaf),
            (final_csaf['document'], 'references', self.document_references.csaf),
            (final_csaf['document'], 'acknowledgments', self.document_acknowledgments.csaf),
            (final_csaf, 'product_tree', self.product_tree.csaf),
            (final_csaf, 'vulnerabilities', self.vulnerability.csaf),
        )
        for target, slot, value in mappings:
            if value:
                target[slot] = value
```

Short example test:
```python
>>> t = {'a': {}}
>>> b = None  # no insertion of b value
>>> c = True
>>> d = 42
>>> e = False  #no insertion of e value
>>> mappings = ((t['a'], 'bkey', b), (t['a'], 'ckey', c), (t, 'dkey', d), (t, 'ekey', e))
>>> for target, slot, value in mappings:
...     if value:
...         target[slot] = value
...
>>> t
{'a': {'ckey': True}, 'dkey': 42}
```

... just saying - all well as it is already.

File: cvrf2csaf/config/config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (sustefil) [reply cvrf2csaf/config/config.yaml]: See:

https://github.com/csaf-tools/CVRF-CSAF-Converter/pull/54/commits/1340080cbad24503ebd88ace7e4fae185871d2f3

Taking the version with every build with `git describe --always` (either annotated tag or commit, if tag is missing).

Then it's accessible via:

- __version__ attribute
- cvrf2csaf -v or --version from CLI

And the version is injected into the CSAF's `document/document_tracking/generator/engine/version`

File: cvrf2csaf/section_handlers/product_tree.py (hunk 1 above)
Comment (sustefil) [reply cvrf2csaf/section_handlers/product_tree.py]: Done, see:

https://github.com/csaf-tools/CVRF-CSAF-Converter/pull/54/commits/6f543cd05603c27172804e03e6c21546e9ec35a9

Comment (sustefil) [reply cvrf2csaf/section_handlers/product_tree.py]: Btw I haven't seen these types in any example so far

File: cvrf2csaf/cvrf2csaf.py (hunk 3 above)
Comment (sustefil) [reply cvrf2csaf/cvrf2csaf.py]: Nice one :) implemented

**Comment by sustefil** [comment]
Quite a bigger fixing batch, @tschmidtb51 please review

File: cvrf2csaf/section_handlers/product_tree.py (hunk 1 above)
Comment (tschmidtb51) [reply cvrf2csaf/section_handlers/product_tree.py]: BTW: This is now also part of the [editor revision](https://github.com/oasis-tcs/csaf/blob/editor-revision-2022-02-23/csaf_2.0/prose/csaf-v2-editor-draft.md#915-conformance-clause-5-cvrf-csaf-converter).

File: cvrf2csaf/config/config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply cvrf2csaf/config/config.yaml]: Sounds perfect. Thanks.

**Review by tschmidtb51 — APPROVED** [review]
BTW: The conversion rule of `Realm` and `Resource` is now part of the [editor revision](https://github.com/oasis-tcs/csaf/blob/editor-revision-2022-02-23/csaf_2.0/prose/csaf-v2-editor-draft.md#915-conformance-clause-5-cvrf-csaf-converter).

**Pull request merged**